src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1021 @@
+use self::scopeguard::guard;
+use alloc::{alloc, dealloc, handle_alloc_error};
+use core::alloc::Layout;
+use core::hint;
+use core::iter::FusedIterator;
+use core::marker::PhantomData;
+use core::mem;
+use core::mem::ManuallyDrop;
+use core::ptr::NonNull;
+
+// Extracted from the scopeguard crate
+mod scopeguard {
+    use core::ops::{Deref, DerefMut};
+    pub struct ScopeGuard<T, F>
+    where
+        F: FnMut(&mut T),
+    {
+        dropfn: F,
+        value: T,
+    }
+    #[inline]
+    pub fn guard<T, F>(value: T, dropfn: F) -> ScopeGuard<T, F>
+    where
+        F: FnMut(&mut T),
+    {
+        ScopeGuard { dropfn, value }
+    }
+    impl<T, F> Deref for ScopeGuard<T, F>
+    where
+        F: FnMut(&mut T),
+    {
+        type Target = T;
+        #[inline]
+        fn deref(&self) -> &T {
+            &self.value
+        }
+    }
+    impl<T, F> DerefMut for ScopeGuard<T, F>
+    where
+        F: FnMut(&mut T),
+    {
+        #[inline]
+        fn deref_mut(&mut self) -> &mut T {
+            &mut self.value
+        }
+    }
+    impl<T, F> Drop for ScopeGuard<T, F>
+    where
+        F: FnMut(&mut T),
+    {
+        #[inline]
+        fn drop(&mut self) {
+            (self.dropfn)(&mut self.value)
+        }
+    }
+}
+
+// Branch prediction hint. This is currently only available on nightly but it
+// consistently improves performance by 10-15%.
+use core::intrinsics::{likely, unlikely};
+
+// Use the SSE2 implementation if possible: it allows us to scan 16 buckets at
+// once instead of 8.
+#[cfg(all(
+    not(stage0),
+    target_feature = "sse2",
+    any(target_arch = "x86", target_arch = "x86_64")
+))]
+#[path = "sse2.rs"]
+mod imp;
```

Comment:
I'd guess that we can't use run-time feature detection here because that is part of `libstd` right? E.g. an AVX implementation would be able to scan 32 buckets at once, and an AVX-512 implementation 64 buckets.

Reply:
The probability of finding a matching bucket drops the further you are from the initial probe location, so there is not much point in scanning more than 16 elements at once. Also, introducing run-time feature detection would require going through function pointers which will demolish performance.

In fact on ARM systems we don't even try to use NEON instructions since they tend to have high latency (usually around 3-5 cycles, whereas SSE instruction are single-cycle). After some experiments I determined that just using the generic integer version which scans 4/8 bytes is faster.

Reply:
> The probability of finding a matching bucket drops the further you are from the initial probe location, so there is not much point in scanning more than 16 elements at once. 

Makes sense, could you add a comment with this information ?